src/charm.py

```diff
@@ -103,7 +117,119 @@ def _on_stop(self, _: ops.StopEvent) -> None:
     def _on_upgrade_charm(self, _: ops.UpgradeCharmEvent) -> None:
         """Handle upgrade-charm."""
         self.unit.status = ops.MaintenanceStatus("Upgrading dependencies")
-        self.bind.prepare(self.unit.name)
+        self.bind.setup(self.unit.name)
+
+    def _on_leader_elected(self, _: ops.LeaderElectedEvent) -> None:
+        """Handle leader-elected event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _on_peer_relation_departed(self, _: ops.RelationDepartedEvent) -> None:
+        """Handle the peer relation departed event."""
+        # We check that we are still the leader when starting to process this event
+        if self.unit.is_leader():
+            self._become_active()
+
+    def _become_active(self) -> bool:
+        """Set the current unit as the active unit of the charm.
+
+        Returns:
+            True if the charm is effectively the new active unit.
+        """
+        active_unit_ip = self._active_unit_ip()
+
+        assert self.peer_relation is not None  # nosec
+        if not active_unit_ip:
+            self.peer_relation.data[self.app].update({"active-unit": self._unit_ip()})
+            return True
+
+        if active_unit_ip != self._unit_ip():
+            status = self._dig_query(
+                f"@{active_unit_ip} service.{constants.ZONE_SERVICE_NAME} TXT +short",
+                retry=True,
+                wait=1,
+            )
+            if status != "ok":
+                self.peer_relation.data[self.app].update({"active-unit": self._unit_ip()})
+                return True
+            return False
+
+        return True
+
+    async def _dig_query(self, cmd: str, retry: bool = False, wait: int = 5) -> str:
```

Comment:
I don't think we can use `async` with charm, i would love to be incorrect here though. Is async now supported?

Reply:
I removed the unused async keyword